tpm2_pytss/types.py

```diff
@@ -732,399 +743,715 @@ class TPMA_MEMORY(int):
 
-class TPMS_CERTIFY_INFO(object):
-    def __new__(self):
-        return ffi.new("TPMS_CERTIFY_INFO *")
 
+class TPMS_KEYEDHASH_PARMS(TPM_OBJECT):
+    def __init__(self, _cdata=None):
+        init_data = TPM_OBJECT._init_cdata(self, _cdata)
+        super().__init__(init_data)
 
-class TPMS_CLOCK_INFO(object):
-    def __new__(self):
-        return ffi.new("TPMS_CLOCK_INFO *")
 
+class TPMS_NV_CERTIFY_INFO(TPM_OBJECT):
+    def __init__(self, _cdata=None):
+        init_data = TPM_OBJECT._init_cdata(self, _cdata)
+        super().__init__(init_data)
 
-class TPMS_COMMAND_AUDIT_INFO(object):
-    def __new__(self):
-        return ffi.new("TPMS_COMMAND_AUDIT_INFO *")
 
+class TPMS_NV_PIN_COUNTER_PARAMETERS(TPM_OBJECT):
+    def __init__(self, _cdata=None):
+        init_data = TPM_OBJECT._init_cdata(self, _cdata)
+        super().__init__(init_data)
 
-class TPMS_CONTEXT(object):
-    def __new__(self):
-        return ffi.new("TPMS_CONTEXT *")
 
+class TPMS_NV_PUBLIC(TPM_OBJECT):
+    def __init__(self, _cdata=None):
+        init_data = TPM_OBJECT._init_cdata(self, _cdata)
+        super().__init__(init_data)
 
-class TPMS_CONTEXT_DATA(object):
-    def __new__(self):
-        return ffi.new("TPMS_CONTEXT_DATA *")
 
+class TPMS_PCR_SELECT(TPM_OBJECT):
+    def __init__(self, _cdata=None):
+        init_data = TPM_OBJECT._init_cdata(self, _cdata)
+        super().__init__(init_data)
 
-class TPMS_CREATION_DATA(object):
-    def __new__(self):
-        return ffi.new("TPMS_CREATION_DATA *")
 
+class TPMS_PCR_SELECTION(TPM_OBJECT):
+    def __init__(self, halg=0, pcrs=None, _cdata=None):
+        init_data = TPM_OBJECT._init_cdata(self, _cdata)
+        super().__init__(init_data)
 
-class TPMS_CREATION_INFO(object):
-    def __new__(self):
-        return ffi.new("TPMS_CREATION_INFO *")
+        if not halg and not pcrs:
+            return
 
+        if bool(halg) != bool(pcrs):
+            raise RuntimeError("halg and pcrs MUST be specified")
 
-class TPMS_ECC_PARMS(object):
-    def __new__(self):
-        return ffi.new("TPMS_ECC_PARMS *")
+        self._cdata.hash = halg
+        self._cdata.sizeofSelect = 3
 
+        if pcrs == "all" or (len(pcrs) == 1 and pcrs[0] == "all"):
+            self._cdata.pcrSelect[0] = 0xFF
+            self._cdata.pcrSelect[1] = 0xFF
+            self._cdata.pcrSelect[2] = 0xFF
+            return
 
-class TPMS_ECC_POINT(object):
-    def __new__(self):
-        return ffi.new("TPMS_ECC_POINT *")
+        for pcr in pcrs:
+            if pcr < 0 or pcr > lib.TPM2_PCR_LAST:
+                raise RuntimeError(f"PCR Index out of range, got {pcr}")
+            self._cdata.pcrSelect[pcr // 8] |= 1 << (pcr % 8)
 
+    @staticmethod
+    def parse(selection):
 
-class TPMS_EMPTY(object):
-    def __new__(self):
-        return ffi.new("TPMS_EMPTY *")
+        if selection is None or len(selection) == 0:
+            return TPMS_PCR_SELECTION()
 
+        hunks = [x.strip() for x in selection.split(":")]
```

Comment:
You might want to check for the presence of ":" before splitting on it, and similarly raise an exception. Users might get an unhelpful error message if they didn't supply a value with a ":".

Reply:
Yep, sounds good, here and throughout on splits from user data

Reply:
Theirs a raise below it, we MUST have 2 chunks. So I can make that exception better.